src/get-ipfs.ts

```diff
@@ -21,125 +26,144 @@ interface IpfsPkg extends ipfs {
-    if (ipfs.enable) {
-      const permissions = normalizePermissions(config.permissions)
-      ipfs = await ipfs.enable({
-        commands: permissions
-      })
-    }
-    const isWorking = await ipfsIsWorking(ipfs)
-    if (isWorking) {
-      return ipfs as ipfs
-    } else {
-      return null
-    }
-  }
-  return null
-}
 
-export const loadIpfsScript = (): Promise<void> => {
-  return new Promise((resolve, reject) => {
-    const script = document.createElement('script')
-    script.src = 'https://unpkg.com/ipfs@0.38.0/dist/index.min.js'
-    script.onload = async () => {
-      return resolve()
-    }
-    script.onerror = reject
-    document.body.appendChild(script)
-  })
+export async function loadJsIpfs(config: config): Promise<ipfs | null> {
+  const ipfsPkg = await (() => { switch (typeof config.jsIpfs) {
+    case "function":  return config.jsIpfs()
+    case "string":    return loadJsIpfsThroughScript(config.jsIpfs)
+    default:          return loadJsIpfsThroughScript(UNPKG_URL)
+  }})()
+
+  if (!ipfsPkg?.create) return null
+  const peers = (config.peers || []).concat(config.browserPeers || [])
+  const ipfs = await ipfsPkg.create({ config: { Addresses: { Swarm: peers }}})
+  return await ipfsIsWorking(ipfs) ? ipfs : null
 }
 
-export const loadJsIpfs = async (config: config): Promise<ipfs | null> => {
-  try {
-    await loadIpfsScript()
-  } catch (err) {
-    return null
-  }
 
-  const ipfsPkg = (window as IpfsWindow).Ipfs
-  if (!ipfsPkg || !ipfsPkg.create) {
-    return null
-  }
-  const ipfs = await ipfsPkg.create()
-  const isWorking = await ipfsIsWorking(ipfs)
+export async function loadWindowIpfs(config: config): Promise<ipfs | null> {
+  let ipfs = (window as IpfsWindow).ipfs
+  if (!ipfs) return null
 
-  if (isWorking) {
-    return ipfs
-  } else {
-    return null
+  if (ipfs.enable) {
+    const permissions = normalizePermissions(config.permissions)
+    ipfs = await ipfs.enable({ commands: permissions })
   }
+
+  return await ipfsIsWorking(ipfs)
+    ? ipfs as ipfs
+    : null
 }
 
-const getIpfs = async (config: config = {}): Promise<ipfs> => {
-  // if already instantiated
-  if (ipfsInstance) {
-    return ipfsInstance
-  }
 
-  let peers = config.peers || []
-  ipfsInstance = await loadWindowIpfs(config)
-  if (ipfsInstance) {
-    console.log('window.ipfs is available!')
-    peers = peers.concat(config.localPeers || [])
-  } else {
-    console.log('window.ipfs is not available, downloading from CDN...')
-    ipfsInstance = await loadJsIpfs(config)
-    peers = peers.concat(config.browserPeers || [])
-  }
 
+// ㊙️
+
+
+async function connectPeers(ipfs: ipfs, peers: string[] = []): Promise<ipfs> {
   if (!ipfsInstance) {
     throw new Error('Could not load IPFS')
   }
 
-  await connectPeers(ipfsInstance, peers)
-  console.log('connected to: ', peers)
+  await Promise.all(
+    peers.map(async p => {
+      try {
+        await ipfs.swarm.connect(p)
+        console.log('Connected to:', peers)
```

Comment:
This is going to get logged for every peer that connects. Would be better to log after attempting all connections.

You could either keep try of the successful/failed connections or you can get the same info with `await ipfs.swarm.peers()` (which is very fast)

Reply:
Done 👍  It now lists all the peers you're connected to:

> Connected to: node1.preload.ipfs.io:443, ams-1.bootstrap.libp2p.io:443, lon-1.bootstrap.libp2p.io:443, nyc-1.bootstrap.libp2p.io:443, nyc-2.bootstrap.libp2p.io:443, sfo-3.bootstrap.libp2p.io:443, node.fission.systems:4003

Reply:
Might have to take another approach, because now using ipfs companion it says:

> Connected to: 108.61.178.31:4001, 111.229.124.163:4001, 111.229.153.141:4001, 111.229.169.146:4001, 111.229.232.65:4001, 111.229.233.109:4001, 111.229.233.190:4001, 111.229.54.203:4001, 111.230.174.8:4001, 122.51.145.92:4001, 122.51.190.189:4001, 123.206.102.92:4001, 125.121.153.191:56800, 134.175.199.83:4001, 134.175.244.169:4001, 138.201.67.220:4001, 138.207.64.87:4001, 139.180.136.130:20001, 139.180.146.71:4001, 144.34.233.111:4001, 147.75.192.141:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.195.153:4001, 147.75.90.11:4001, 157.230.173.31:4001, 158.101.170.135:4001, 159.89.145.27:4001, 165.227.32.248:4001, 167.99.132.247:4001, 175.24.32.138:4001, 18.130.148.15:4001, 185.18.6.160:4001, 192.99.9.35:4001, 198.100.145.82:4001, 206.116.153.42:4003, 206.189.164.232:4001, 207.148.19.196:20027, 207.148.19.196:20035, 207.148.19.196:20045, 23.111.231.167:4001, 3.215.160.238:4001, 31.40.53.200:65242, 34.83.208.153:4001, 50.7.177.19:4001, 52.47.147.55:4001, 52.56.144.76:20009, 52.56.144.76:20026, 52.56.144.76:20029, 52.56.144.76:20030, 52.56.144.76:20051, 52.56.144.76:20060, 66.42.111.225:20003, 66.42.111.225:20019, 66.42.111.225:20037, 66.42.111.225:20049, 66.42.111.225:20050, 66.42.111.225:20055, 66.42.111.225:20066, 70.29.76.231:30997, 80.184.100.35:38814, 94.23.206.224:4001, 97.80.184.175:62534, 2409:8a00:6c8d:51f0:6866:af79:206f:4947:4001

A bit much 😅

Reply:
haha yeah maybe just logging the array object so that it gets automatically collapsed